psamm/commands/completepath.py

```diff
@@ -0,0 +1,280 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# PSAMM is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with PSAMM.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Copyright 2014-2017  Jon Lund Steffensen <jon_steffensen@uri.edu>
+
+from __future__ import unicode_literals
+
+import logging
+import argparse
+
+from six import text_type, iteritems
+from ..reaction import Reaction, Direction
+from ..command import (Command, MetabolicMixin, SolverCommandMixin,
+                       FilePrefixAppendAction)
+from ..gapfill import gapfill, GapFillError
+from ..datasource.reaction import parse_compound
+from ..gapfilling import create_extended_sink_source_model
+from ..lpsolver import lp
+
+logger = logging.getLogger(__name__)
+
+
+class CompletePathCommand(MetabolicMixin, SolverCommandMixin, Command):
+    """Run the CompletePath method on the model."""
+    @classmethod
+    def init_parser(cls, parser):
+        parser.add_argument(
+            '--compound', metavar='compound', action=FilePrefixAppendAction,
+            type=parse_compound, default=[],
+            help='Select compounds to try to unblock')
+        parser.add_argument(
+            '--penalty', metavar='file', type=argparse.FileType('r'),
+            help='List of penalty scores for database reactions')
+        parser.add_argument(
+            '--db-penalty', metavar='penalty', type=float,
+            help='Default penalty for database reactions')
+        parser.add_argument(
+            '--tp-penalty', metavar='penalty', type=float,
+            help='Default penalty for transport reactions')
+        parser.add_argument(
+            '--source-penalty', metavar='penalty', type=float,
+            help='Default penalty for exchange reactions')
+        parser.add_argument(
+            '--sink-penalty', metavar='penalty', type=float,
+            help='Default penalty for exchange reactions')
+        parser.add_argument(
+            '--epsilon', type=float, default=1e-5,
+            help='Threshold for reaction flux')
+        parser.add_argument(
+            '--implicit-sinks', action='store_true',
+            help='Do not include implicit sinks when gap-filling')
+        parser.add_argument(
+            '--allow-bounds-expansion', action='store_true',
+            help=('Allow GapFill to propose expansion of flux bounds. This'
+                  ' includes turning irreversible reactions reversible.'))
+        parser.add_argument(
+            '--gapfill', action='store_true',
+            help='Run the gapfill algorithm on the model with separated '
+                 'sink and source reactions in each compatment.')
```

Comment:
This would be running the gapfill but with the modified extended model that includes separate sources/sinks in every model compartment. So the results will be slightly different than the other gapfill implementation.